care/emr/api/viewsets/valueset.py

```diff
@@ -98,3 +107,100 @@ def lookup_code(self, request, *args, **kwargs):
                 {"error": "No results found for the given system and code"}, status=404
             )
         return Response(result)
+
+    def _get_or_create_user_preferences(self, user, valueset) -> UserValueSetPreference:
+        return UserValueSetPreference.objects.get_or_create(
+            user=user,
+            valueset=valueset,
+            defaults={"favorite_codes": []},
+        )[0]
+
+    @extend_schema(request=MinimalCodeConcept, responses={200: None}, methods=["POST"])
+    @action(detail=True, methods=["POST"])
+    def add_favourite(self, request, *args, **kwargs):
+        valueset = self.get_object()
+        code_obj = MinimalCodeConcept(**request.data)
+
+        # validate the code
+        if not valueset.lookup(code_obj):
+            raise ValidationError("Invalid value")
+
+        preferences = self._get_or_create_user_preferences(request.user, valueset)
+
+        if len(preferences.get_favourites()) >= preferences.MAX_FAVORITES:
+            raise ValidationError("Maximum number of favorites reached (50)")
+
+        preferences.add_favourite(code_obj.model_dump())
+        preferences.add_recent_view(
+            code_obj.model_dump()
+        )  # Add to recent views as well
+        return Response({"message": f"Code {code_obj.code} marked as favorite"})
+
+    @action(detail=True, methods=["POST"])
+    def remove_favourite(self, request, *args, **kwargs):
+        valueset = self.get_object()
+        code_obj = MinimalCodeConcept(**request.data)
+
+        preferences = self._get_or_create_user_preferences(request.user, valueset)
+        preferences.remove_favourite(code_obj.code)
+        return Response({"message": f"Code {code_obj.code} removed from favorites"})
+
+    @action(detail=True, methods=["POST"])
+    def clear_favourites(self, request, *args, **kwargs):
+        valueset = self.get_object()
+        preference = UserValueSetPreference.objects.filter(
+            user=request.user, valueset=valueset
+        ).first()
+        if preference:
+            preference.clear_favourites()
+        return Response({"message": "All favorite codes cleared"})
+
+    @action(detail=True, methods=["GET"])
+    def favourites(self, request, *args, **kwargs):
+        valueset = self.get_object()
+        preferences = UserValueSetPreference.objects.filter(
+            user=request.user, valueset=valueset
+        ).first()
+        return Response(preferences.get_favourites() if preferences else [])
+
+    @extend_schema(request=MinimalCodeConcept, responses={200: None}, methods=["POST"])
+    @action(detail=True, methods=["POST"])
+    def add_recent_view(self, request, *args, **kwargs):
+        valueset = self.get_object()
+        code_obj = MinimalCodeConcept(**request.data)
+
+        # validate the code
+        if not valueset.lookup(code_obj):
+            raise ValidationError("Invalid value")
+
+        preferences = self._get_or_create_user_preferences(request.user, valueset)
```

Comment:
Can we make this static, no need for the db call. for any of the recent view bits.